src/apps/services/cache.py

```diff
@@ -0,0 +1,281 @@
+
+    def date_agg(self, date_column, freq):
+        if freq == "daily":
+            return self.daily_agg(date_column)
+        elif freq == "monthly":
+            return self.monthly_agg(date_column)
+        else:
+            raise DashArgumentException("Unknown date aggregation frequency")
+
+    @chained_cached_command()
+    def daily_agg(self, df, columns):
+        if not isinstance(columns, list):
+            columns = [columns]
+        date_column = columns[0]
+        """Adds an aggregation by day"""
+        df = self.date_manipulations(df, date_column, "daily")
+        df = df.groupby(columns)
+        return df
+
+    @chained_cached_command()
+    def monthly_agg(self, df, columns):
+        """Adds an aggregation by month"""
+        if not isinstance(columns, list):
+            columns = [columns]
+        date_column = columns[0]
+        df = self.date_manipulations(df, date_column, "monthly")
+        df = df.groupby(columns)
+        return df
+
+    @final_cached_command()
+    def agg(self, df, column, operators):
+        """Sums results, works also on aggregations"""
+        res = (
+            df[column]
+            .agg(list(operators.keys()))
+            .reset_index()
+            .rename(columns=operators)
+        )
+        return res
+
+    @final_cached_command()
+    def sum(self, df, column):
+        """Sums results, works also on aggregations"""
+        res = df[column].sum()
+
+        # Reset index if we are computing aggregated sums
+        if isinstance(res, pd.core.series.Series):
+            res = res.reset_index()
+
+        return res
+
+    @final_cached_command()
+    def sum_over_time(self, df, date_column, column, freq):
+        df = self.date_manipulations(df, date_column, freq)
+        df = df.sort_values(by=date_column, ascending=True)
+        df = df.groupby(date_column)[column].sum().to_frame().reset_index()
+        df[column] = df[column].cumsum()
+        return df
+
+    @final_cached_command()
+    def cumsum(self, df, column):
+        """Cumulative sum"""
+        return df[column].cumsum()
+
+    def date_manipulations(self, df, date_column, freq):
+        if date_column not in df:
+            raise DashArgumentException(f"Unknown column '{date_column}'")
+        if freq == "daily":
+            return self.date_manipulations_daily(df, date_column)
+        elif freq == "monthly":
+            return self.date_manipulations_monthly(df, date_column)
+
+    def date_manipulations_monthly(self, df, date_column):
+        df[date_column] = pd.to_datetime(df[date_column]).dt.to_period("M")
+        df[date_column] = pd.to_datetime(df[date_column].dt.start_time).dt.date
+        return df
+
+    def date_manipulations_daily(self, df, date_column):
+        if not (df.empty):
+            df[date_column] = (
+                df[date_column]
+                .dt.tz_localize(None)
+                .dt.floor("D")
+                .dt.date
+            )
+            """
```

Comment:
What's the deal with this docstring comment? Unnecessary right now? More functionality to be added later?

Should be a `TODO:` on it if so...

Reply:
Ho no my bad. I just forgot to remove it.